p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,492 @@ where
+            return Err(P2pError::ProtocolError(ProtocolError::DisconnectedHeaders));
         }
 
-        // call chainstate to get the blocks that the local node doesn't know about
-        match self
+        // The first header must be connected to a known block.
+        let prev_id = *headers
+            .first()
+            // This is OK because of the `headers.is_empty()` check above.
+            .expect("Headers shouldn't be empty")
+            .prev_block_id();
+        if self
             .chainstate_handle
-            .call(|this| this.filter_already_existing_blocks(headers))
-            .await?
+            .call(move |c| c.get_gen_block_index(&prev_id))
+            .await??
+            .is_none()
         {
-            Ok(headers) => {
-                peer.register_header_response(&headers);
-                Ok(peer.get_header_for_download())
-            }
-            Err(err) => Err(P2pError::ChainstateError(err)),
+            return Err(P2pError::ProtocolError(ProtocolError::DisconnectedHeaders));
         }
-    }
 
-    /// Process incoming header response
-    pub async fn process_header_response(
-        &mut self,
-        peer_id: T::PeerId,
-        headers: Vec<BlockHeader>,
-    ) -> crate::Result<()> {
-        match self.validate_header_response(&peer_id, headers).await {
-            Ok(Some(header)) => self.send_block_request(peer_id, header.get_id()),
-            Ok(None) => {
-                self.peers
-                    .get_mut(&peer_id)
-                    .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?
-                    .set_state(peer::PeerSyncState::Idle);
-                Ok(())
+        let is_max_headers =
+            headers.len() == Into::<usize>::into(self.p2p_config.header_limit.clone());
+        let headers = self
+            .chainstate_handle
+            .call(|c| c.filter_already_existing_blocks(headers))
+            .await??;
+        if headers.is_empty() {
+            // A peer can have more headers if we have received the maximum amount of them.
+            if is_max_headers {
+                self.request_headers(peer).await?;
             }
-            Err(err) => Err(err),
+            return Ok(());
         }
+
+        // Only the first header can be checked with the `preliminary_header_check` function.
+        let first_header = headers
+            .first()
+            // This is OK because of the `headers.is_empty()` check above.
+            .expect("Headers shouldn't be empty")
+            .clone();
+        self.chainstate_handle
+            .call(|c| c.preliminary_header_check(first_header))
+            .await??;
+
+        self.request_blocks(peer, headers)
     }
 
-    /// Validate incoming block response
-    async fn validate_block_response(
-        &mut self,
-        peer_id: &T::PeerId,
-        blocks: Vec<Block>,
-    ) -> crate::Result<Option<BlockHeader>> {
-        let peer = self
+    // TODO: This shouldn't be public.
```

Comment:
Unfortunately, it is an artifact from the previous design where almost everything was made public for tests. I wasn't sure that I would be able to fix everything in this pull request, but currently I'm trying to rewrite the tests and hopefully this can be removed too.

Reply:
Most of the methods become private now, but unfortunately not all of them.